org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexer.java

```diff
@@ -271,4 +284,36 @@ public void indexResolvedDocument() {
 			}
 		}
 	}
+
+	/**
+	 * @return whether the operatin was successful
+	 */
+	private boolean indexDocumentFromDOM() {
+		if (this.document instanceof JavaSearchDocument javaSearchDoc) {
+			IFile file = javaSearchDoc.getFile();
+			try {
+				if (file.getProject().getNature(JavaCore.NATURE_ID) instanceof IJavaProject javaProject) {
+					IClasspathEntry cpEntry = javaProject.findContainingClasspathEntry(file);
+					IJavaElement element = javaProject.findElement(file.getFullPath().makeRelativeTo(cpEntry.getPath()));
+					if (element instanceof org.eclipse.jdt.internal.core.CompilationUnit modelUnit) {
+						// TODO check element info: if has AST and flags are set sufficiently, just reuse instead of rebuilding
+						ASTParser astParser = ASTParser.newParser(modelUnit.getElementInfo() instanceof ASTHolderCUInfo astHolder ? astHolder.astLevel : AST.getJLSLatest());
+						astParser.setSource(modelUnit);
+						astParser.setResolveBindings(false);
+						astParser.setProject(javaProject);
+						astParser.setIgnoreMethodBodies(true);
+						org.eclipse.jdt.core.dom.ASTNode dom = astParser.createAST(null);
+						if (dom != null) {
+							dom.accept(new DOMToIndexVisitor(this));
+							return true;
+						}
+					}
+				}
+			} catch (Exception ex) {
+				ILog.get().error(ex.getMessage(), ex);
```

Comment:
ex.getMessage() can be null for example during NPE. better provide a significant message including  the file name.

Reply:
OK, will do.